src/mesh_tools.rs

```diff
@@ -5,8 +5,161 @@ use nalgebra::geometry::Point3;
+    let mut unoriented_edge_index_map: HashMap<UnorientedEdge, u32> =
+        HashMap::with_capacity(unoriented_edges.len());
+    for (unoriented_edge_index, unoriented_edge) in unoriented_edges.iter().enumerate() {
+        unoriented_edge_index_map.insert(*unoriented_edge, cast_u32(unoriented_edge_index));
+    }
+
+    // Synchronize faces in all mesh geometry triangles
+    while synchronized_triangle_faces.len() < original_triangle_faces.len() {
+        // Start with the first untreated face in the list of the original mesh
+        // geometry faces. The winding of this face also determines the winding
+        // of the rest of the mesh geometry.
+        let mut current_face_index = cast_u32(
+            face_processed_pattern
+                .iter()
+                .position(|&v| !v)
+                .expect("Face index not found"),
+        );
+
+        face_processed_pattern[cast_usize(current_face_index)] = true;
+        let mut current_test_edge: OrientedEdge =
+            original_triangle_faces[cast_usize(current_face_index)].to_oriented_edges()[0];
+
+        // Put the first edge-face couple to the stack
+        edge_face_stack.push((current_test_edge, current_face_index));
+
+        // Check and revert (if needed) faces in the stack as long as there is any
+        while !edge_face_stack.is_empty() {
+            // Get ready for the next iteration.
+            let next_edge_face = edge_face_stack.pop().expect("Popping from an empty vector");
+            current_test_edge = next_edge_face.0;
+            current_face_index = next_edge_face.1;
+
+            let current_face = original_triangle_faces[cast_usize(current_face_index)];
+            // Use if contains reverted edge, if not, revert and use. It should be
+            // safe to presume the face contains the edge one way or another because
+            // it's given by the topology generator.
+            let proper_current_face = if current_face.contains_oriented_edge(current_test_edge) {
+                current_face
+            } else {
+                current_face.to_reverted()
+            };
+            synchronized_triangle_faces.push(proper_current_face);
+
+            // Find the indices of edges of the current face in the list of edges,
+            // from which the topology was created
+            let face_unoriented_edges = proper_current_face.to_unoriented_edges();
+            let face_unoriented_edge_indices =
+                face_unoriented_edges.iter().map(|unoriented_edge| {
+                    unoriented_edge_index_map
+                        .get(unoriented_edge)
+                        .expect("The current edge is not found in the edge collection")
+                });
+
+            // Convert into oriented edges so that the neighboring faces can check
+            // for correct winding
+            let face_oriented_edges: Vec<_> =
+                face_unoriented_edges.iter().map(|u_e| u_e.0).collect();
```

Comment:
This seems unnecessary. Can't the oriented edge just be extracted of the unoriented edge on demand in the loop below? You can use `Iterator::zip` make it available.

Reply:
It's possible that some of the unoriented edges will be unwrapped more than once. Maybe it's not a performance issue, but that was my original reasoning for this solution.

Reply:
kk, should actually be a perf benefit